### PR TITLE
CODETOOLS-7903176: Move LJV inside JOL

### PR DESCRIPTION
**Pull request by inponomarev**
This moves https://github.com/atp-mipt/ljv project inside JOL in order for LJV to get access to internal JOL's APIs.

See blogposts about LJV [in Russian](https://habr.com/ru/post/599045/) /  [in English](https://dzone.com/articles/what-can-we-learn-from-java-data-structures-visual).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903176](https://bugs.openjdk.org/browse/CODETOOLS-7903176): Move LJV inside JOL (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer)
 * @JohnHamer-Glasgow (no known openjdk.org user name / role) ⚠️ Review applies to [4d44d66e](https://git.openjdk.org/jol/pull/24/files/4d44d66ee3e534624c4b5888a4e6f11e771762c1)

### Contributors
 * Ivan Arkhipov `<me@endevir.ru>`
 * John Hamer `<john.hamer@glasgow.ac.uk>`
 * Ilya Selivanau `<ilyaspacemail@gmail.com>`
 * Victor Petrosyan `<victor@galahad.ee>`
 * Georgy Guminov `<georgy@galahad.ee>`
 * Ilya Shvetsov `<ilya@galahad.ee>`
 * Nuras Nogaev `<nuras@galahad.ee>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jol.git pull/24/head:pull/24` \
`$ git checkout pull/24`

Update a local copy of the PR: \
`$ git checkout pull/24` \
`$ git pull https://git.openjdk.org/jol.git pull/24/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24`

View PR using the GUI difftool: \
`$ git pr show -t 24`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jol/pull/24.diff">https://git.openjdk.org/jol/pull/24.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jol/pull/24#issuecomment-1117819796)

### CODE REVIEW

**Comment by shipilev** [comment]
Thanks for contribution!

Before continuing with code review, there is some high-level things involved in this. I see there are multiple contributors for the incoming changesets. I believe all of them should have signed OCA on file, or their changesets should not be present in this PR (I see some were adding "slides" that are not present in this PR). Right, @robilad? 

You can use `/contributor add <github-id>` to mark them as contributors, I'd expect OpenJDK bots would then block the PR until OCAs are present.

**Comment by inponomarev** [comment]
> I believe all of them should have signed OCA
I will check with everyone personally, and as soon as they confirm they signed OCA, I will add them using `/contributor add`

**Comment by erikj79** [comment]
The Skara bot is currently experiencing trouble processing this PR. Jcheck is throwing an exception. I'm still investigating. The reason seems to be that this this is merging in something that does not have a merge-base with the master branch. It's possible that you won't be able to integrate before I resolve this. Filed https://bugs.openjdk.java.net/browse/SKARA-1424.

**Comment by robilad** [comment]
Aleksey is correct that we need to have an OCA on file for each contributor involved.

**Comment by robilad** [comment]
As another note, it's not sufficient that an OCA has been signed - it has to successfully be processed on our end, which means it shows up on the signatory listing at oca.opensource.Oracle.com.

File: jol-core/src/main/java/org/openjdk/jol/ljv/Introspection.java
Comment (shipilev) [inline jol-core/src/main/java/org/openjdk/jol/ljv/Introspection.java:55]: Here and later: newlines at the end of files.

**Review by shipilev — CHANGES_REQUESTED** [review]
I see there are some OCA signatories, but not all contributors. Please make sure all are present here: https://oca.opensource.oracle.com/?ojr=contrib-list

Cursory review meanwhile:

**Comment by inponomarev** [comment]
Hi @shipilev ! 

It seems that we managed to sort out everything with the OCA 

The code is linted and IMO ready for review

**Comment by shipilev** [comment]
> It seems that we managed to sort out everything with the OCA

Nice, I checked the OCA signatories list, and all PR contributors seems to be accounted for.

In this PR, I see the commits from two more people (@[OrangeFlag](https://github.com/atp-mipt/jol/commits?author=OrangeFlag) and @[gogagum](https://github.com/atp-mipt/jol/commits?author=gogagum) -- and who are not in PR contributor list. I see they have OCA on file, so that is not a problem for this PR, but maybe you want to credit them as well?

> The code is linted and IMO ready for review

I'll start reviewing it then! Thanks.

File: jol-core/src/test/java/org/openjdk/jol/ljv/ArrayItem.java
Comment (shipilev) [inline jol-core/src/test/java/org/openjdk/jol/ljv/ArrayItem.java:1]: Here and later, files are missing copyright headers.

**Review by shipilev — CHANGES_REQUESTED** [review]
Another round of review follows.

File: jol-core/src/test/java/org/openjdk/jol/ljv/HashCodeCollision.java
Comment (inponomarev) [reply jol-core/src/test/java/org/openjdk/jol/ljv/HashCodeCollision.java:9]: reimplemented the class from scratch

File: jol-core/src/main/java/org/openjdk/jol/ljv/GraphvizVisualization.java
Comment (inponomarev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/GraphvizVisualization.java:55]: No it won't be a problem for Approvals tests, as far as I understand it can treat the line ending discrepancies correctly.
However I can notice `\n` in `OptionFormatter` and `ClassLayout`, so maybe postpone the line ending problem?

**Comment by inponomarev** [comment]
Hi @shipilev review comments are addressed, please check when you have time. Thank you!

File: jol-core/src/main/java/org/openjdk/jol/ljv/GraphvizVisualization.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/GraphvizVisualization.java:55]: Okay, sounds fine.

File: jol-core/src/main/java/org/openjdk/jol/ljv/Introspection.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/Introspection.java:32]: This comment is still not addressed.

File: jol-core/src/main/java/org/openjdk/jol/ljv/IntrospectionWithReflectionAPI.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/IntrospectionWithReflectionAPI.java:53]: This comment is still not addressed.

File: jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java:44]: This comment is still not addressed.

File: jol-core/src/main/java/org/openjdk/jol/ljv/Quote.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/Quote.java:53]: This comment was not addressed.

File: jol-core/src/main/java/org/openjdk/jol/ljv/pom.xml
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/pom.xml:1]: This comment was not addressed.

File: jol-core/src/main/java/org/openjdk/jol/ljv/provider/ArrayElementAttributeProvider.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/provider/ArrayElementAttributeProvider.java:27]: This comment was not addressed.

**Review by shipilev — CHANGES_REQUESTED** [review]
There is a bunch of comments that was not addressed since last review, unfortunately.

**Comment by inponomarev** [comment]
@shipilev sorry for the inattention, hope everything is addressed now

File: jol-core/src/main/java/org/openjdk/jol/ljv/Introspection.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/Introspection.java:32]: All right. This would be fine if we have enough test to cover this refactoring. I might get a stab at it after the integration.

File: jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java]: I cannot advise on what you should do here. I see the current PR already does what you suggest, which I presume what John approves? Anyhow, I think we need to ask @robilad for guidance here.

Comment (inponomarev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java]: Yes, John approves indeed. He's not at GitHub, but he's available via email and he approves the whole idea and he's glad that his idea revived.

Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java]: Idea: can John send the reply to https://mail.openjdk.org/pipermail/jol-dev/2022-September/000464.html then?

**Review by JohnHamer-Glasgow — APPROVED** [review]

**Review by JohnHamer-Glasgow — APPROVED** [review]

File: jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/LJV.java]: Thank you for chiming in, @JohnHamer-Glasgow! (And welcome to JOL!)

I have a suggestion to remove `Author` and `Created` lines from this file, because it does not match the style of other files in the project, and this might be lost by accident, or otherwise raise questions during the review. For example, if we ever divide/delete `LJV.java`, it would be hard to judge where these comments should go.

There is a tangential rule that we try to avoid `@author` javadoc tags, to be consistent with OpenJDK code style. I think John is now recorded as developer in project properties, which is cleaner, better, and gives John the credit for the entire project, not just single file.

File: jol-core/src/main/java/org/openjdk/jol/ljv/pom.xml
Comment (inponomarev) [reply jol-core/src/main/java/org/openjdk/jol/ljv/pom.xml]: This POM file was a lefover of the LJV project and it is not included in the final version

**Review by shipilev — APPROVED** [review]
Looks good.

As stated earlier, I'd like to have @robilad's blessing for this large contribution (even though it is not required).

**Comment by shipilev** [comment]
One more thing, @inponomarev -- please merge from current JOL master to get test pipeline fixes. GitHub actions need to test this PR, maybe you need to go to your fork "Actions" tab and press the "Approve workflow" button there.

**Comment by robilad** [comment]
I can confirm that the individuals listed as contributors have an OCA on file.

**Comment by robilad** [comment]
Please also have each contributor listed in the change confirm in the PR comments that they would like to make this contribution, as this contribution has not been developed within this project.

**Comment by Endevir** [comment]
I confirm that I'll be glad to make this contribution within JOL project.

**Comment by VityaP** [comment]
Hello @robilad ,  as a contributor to LJV, I confirm that we wanted to merge it with OpenJDK/JOL

**Comment by OrionChocoPie** [comment]
I confirm that I would like to make this contribution

**Comment by IlyaShvetsov** [comment]
Hello! I confirm that I want this to be merged with OpenJDK/JOL

**Comment by jham005** [comment]
I confirm that I would like this merged.

**Comment by gogagum** [comment]
> Please also have each contributor listed in the change confirm in the PR comments that they would like to make this contribution, as this contribution has not been developed within this project.

I confirm that I want this contribution to be made within JOL project.

**Comment by OrangeFlag** [comment]
> Please also have each contributor listed in the change confirm in the PR comments that they would like to make this contribution, as this contribution has not been developed within this project.

I confirm that I would like to make this contribution

**Pull request closed without merging**
